File: typescript/tests/http.test.ts
import { describe, it, expect, vi } from "vitest";
import { PagouHttpClient } from "../src/lib/http.js";
import { NotFoundError, ServerError, NetworkError } from "../src/lib/errors.js";
import type { PagouConfig } from "../src/lib/config.js";

const config: PagouConfig = {
  environment: "sandbox",
  baseUrl: "https://api.sandbox.pagou.ai",
  apiToken: "test_token",
  timeoutMs: 1000,
  maxRetries: 1,
};

function jsonResponse(status: number, body: unknown, headers: Record<string, string> = {}): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { "content-type": "application/json", ...headers },
  });
}

describe("PagouHttpClient", () => {
  it("unwraps the { success, requestId, data } envelope", async () => {
    const fetchImpl = vi.fn(async () => jsonResponse(200, { success: true, requestId: "req_1", data: { id: "tx_1" } }));
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);

    const { data, requestId } = await client.requestData<{ id: string }>({ method: "GET", path: "/v2/transactions/tx_1" });
    expect(data.id).toBe("tx_1");
    expect(requestId).toBe("req_1");
  });

  it("sends Authorization and a generated correlation id", async () => {
    const fetchImpl = vi.fn(async () => jsonResponse(200, { success: true, requestId: "r", data: {} }));
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);
    await client.requestData({ method: "GET", path: "/v2/transactions" });

    const [, init] = fetchImpl.mock.calls[0] as unknown as [URL, RequestInit];
    const headers = init.headers as Headers;
    expect(headers.get("Authorization")).toBe("Bearer test_token");
    expect(headers.get("X-Request-Id")).toMatch(/[0-9a-f-]{36}/);
  });

  it("maps a 404 to NotFoundError without retrying", async () => {
    const fetchImpl = vi.fn(async () => jsonResponse(404, { message: "not found", code: "NOT_FOUND" }));
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);

    await expect(client.requestData({ method: "GET", path: "/v2/transactions/x" })).rejects.toBeInstanceOf(NotFoundError);
    expect(fetchImpl).toHaveBeenCalledTimes(1);
  });

  it("retries a 500 on GET, then succeeds", async () => {
    const fetchImpl = vi
      .fn()
      .mockResolvedValueOnce(jsonResponse(500, { message: "boom" }))
      .mockResolvedValueOnce(jsonResponse(200, { success: true, requestId: "r", data: { ok: true } }));
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);

    const { data } = await client.requestData<{ ok: boolean }>({ method: "GET", path: "/v2/transactions" });
    expect(data.ok).toBe(true);
    expect(fetchImpl).toHaveBeenCalledTimes(2);
  });

  it("does not retry a POST without an idempotency key", async () => {
    const fetchImpl = vi.fn(async () => jsonResponse(500, { message: "boom" }));
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);

    await expect(client.requestData({ method: "POST", path: "/v2/transactions", body: {} })).rejects.toBeInstanceOf(ServerError);
    expect(fetchImpl).toHaveBeenCalledTimes(1);
  });

  it("retries a POST when an idempotency key is present", async () => {
    const fetchImpl = vi
      .fn()
      .mockResolvedValueOnce(jsonResponse(503, { message: "unavailable" }))
      .mockResolvedValueOnce(jsonResponse(200, { success: true, requestId: "r", data: { id: "tx" } }));
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);

    const { data } = await client.requestData<{ id: string }>({
      method: "POST",
      path: "/v2/transactions",
      body: {},
      idempotencyKey: "idem_1",
    });
    expect(data.id).toBe("tx");
    const [, init] = fetchImpl.mock.calls[0] as unknown as [URL, RequestInit];
    expect((init.headers as Headers).get("Idempotency-Key")).toBe("idem_1");
  });

  it("raises NetworkError with a timeout message when the request aborts", async () => {
    const hangingFetch = (_url: URL, init: RequestInit) =>
      new Promise<Response>((_resolve, reject) => {
        init.signal?.addEventListener("abort", () => {
          const err = new Error("aborted");
          err.name = "AbortError";
          reject(err);
        });
      });
    const client = new PagouHttpClient({ ...config, timeoutMs: 30, maxRetries: 0 }, hangingFetch as unknown as typeof fetch);

    const error = await client.requestData({ method: "GET", path: "/v2/transactions" }).catch((e) => e);
    expect(error).toBeInstanceOf(NetworkError);
    expect((error as NetworkError).message).toBe("Request timed out");
  });

  it("serializes array query params as comma-joined values", async () => {
    const fetchImpl = vi.fn(async () => jsonResponse(200, { success: true, requestId: "r", data: [], next_cursor: null, prev_cursor: null, total: 0 }));
    const client = new PagouHttpClient(config, fetchImpl as unknown as typeof fetch);
    await client.requestCursorPage({ method: "GET", path: "/v2/transactions", query: { paymentMethods: ["pix", "credit_card"] } });

    const [url] = fetchImpl.mock.calls[0] as unknown as [URL, RequestInit];
    expect(url.searchParams.get("paymentMethods")).toBe("pix,credit_card");
  });
});

File: ruby/tests/http_test.rb
# frozen_string_literal: true

require_relative "test_helper"

class HttpClientTest < Minitest::Test
  def test_unwraps_the_envelope
    client = StubClient.new(FakeResponse.new(200,
                                             { "success" => true, "requestId" => "req_1",
                                               "data" => { "id" => "tx_1" }, }))
    result = client.request_data(method: "GET", path: "/v2/transactions/tx_1")
    assert_equal "tx_1", result.data["id"]
    assert_equal "req_1", result.request_id
  end

  def test_sends_authorization_and_generated_correlation_id
    client = StubClient.new(FakeResponse.new(200, { "success" => true, "requestId" => "r", "data" => {} }))
    client.request_data(method: "GET", path: "/v2/transactions")
    req = client.requests.first[:request]
    assert_equal "Bearer test_token", req["Authorization"]
    assert_match(/[0-9a-f-]{36}/, req["X-Request-Id"])
  end

  def test_maps_404_to_not_found_without_retrying
    client = StubClient.new(FakeResponse.new(404, { "message" => "not found", "code" => "NOT_FOUND" }))
    assert_raises(Pagou::NotFoundError) do
      client.request_data(method: "GET", path: "/v2/transactions/x")
    end
    assert_equal 1, client.requests.length
  end

  def test_retries_a_500_on_get_then_succeeds
    client = StubClient.new([
                              FakeResponse.new(500, { "message" => "boom" }),
                              FakeResponse.new(200,
                                               { "success" => true, "requestId" => "r",
                                                 "data" => { "ok" => true }, }),
                            ])
    result = client.request_data(method: "GET", path: "/v2/transactions")
    assert_equal true, result.data["ok"]
    assert_equal 2, client.requests.length
  end

  def test_does_not_retry_a_post_without_idempotency_key
    client = StubClient.new(FakeResponse.new(500, { "message" => "boom" }))
    assert_raises(Pagou::ServerError) do
      client.request_data(method: "POST", path: "/v2/transactions", body: {})
    end
    assert_equal 1, client.requests.length
  end

  def test_retries_a_post_when_an_idempotency_key_is_present
    client = StubClient.new([
                              FakeResponse.new(503, { "message" => "unavailable" }),
                              FakeResponse.new(200,
                                               { "success" => true, "requestId" => "r",
                                                 "data" => { "id" => "tx" }, }),
                            ])
    result = client.request_data(method: "POST", path: "/v2/transactions", body: {},
                                 idempotency_key: "idem_1",)
    assert_equal "tx", result.data["id"]
    assert_equal "idem_1", client.requests.first[:request]["Idempotency-Key"]
  end

  def test_raises_network_error_with_timeout_message_when_the_request_aborts
    client = StubClient.new([Net::ReadTimeout.new], config: TEST_CONFIG.dup.tap { |c| c.max_retries = 0 })
    error = assert_raises(Pagou::NetworkError) do
      client.request_data(method: "GET", path: "/v2/transactions")
    end
    assert_equal "Request timed out", error.message
  end

  def test_serializes_array_query_params_as_comma_joined_values
    client = StubClient.new(FakeResponse.new(200,
                                             { "success" => true, "requestId" => "r", "data" => [], "next_cursor" => nil, "prev_cursor" => nil,
                                               "total" => 0, }))
    client.request_cursor_page(method: "GET", path: "/v2/transactions",
                               query: { "paymentMethods" => %w[pix credit_card] },)
    uri = client.requests.first[:uri]
    params = URI.decode_www_form(uri.query).to_h
    assert_equal "pix,credit_card", params["paymentMethods"]
  end
end
